openff/system/types.py

```diff
@@ -1,6 +1,146 @@
+                # return val
+            elif isinstance(val, (float, int)):
+                return val * unit_
+            elif isinstance(val, str):
+                # could do custom deserialization here?
+                return unit.Quantity(val).to(unit_)
+
+
+class QuantityEncoder(json.JSONEncoder):
+    """JSON encoder for unit-wrapped floats and np arrays. Should work
+    for both FloatQuantity and ArrayQuantity"""
+
+    def default(self, obj):
+        if isinstance(obj, unit.Quantity):
+            if isinstance(obj.magnitude, (float, int)):
+                data = obj.magnitude
+            elif isinstance(obj.magnitude, np.ndarray):
+                data = obj.magnitude.tolist()
+            else:
+                # This shouldn't ever be hit if our object models
+                # behave in ways we expect?
+                raise Exception(
+                    f"trying to serialize unsupported type {type(obj.magnitude)}"
+                )
+            return {
+                "val": data,
+                "unit": str(obj.units),
+            }
+        return json.JSONEncoder.default(self, obj)
+
+
+def custom_quantity_encoder(v):
+    return json.dumps(v, cls=QuantityEncoder)
+
+
+def json_loader(data):
+    data = json.loads(data)
+    for key, val in data.items():
+        try:
+            v = json.loads(val)
+            if "unit" in v:
+                unit_ = unit(v["unit"])
+                val = v["val"]
+                data[key] = unit_ * val
+        except Exception as e:
+            # TODO: Figure out what cases to catch here
+            raise e
+    return data
+
+
+class _ArrayQuantityMeta(type):
+    def __getitem__(self, t):
+        return type("ArrayQuantity", (ArrayQuantity,), {"__unit__": t})
+
+
+class ArrayQuantity(float, metaclass=_ArrayQuantityMeta):
+    @classmethod
+    def __get_validators__(cls):
+        yield cls.validate_type
+
+    @classmethod
+    def validate_type(cls, val):
+        unit_ = getattr(cls, "__unit__", Any)
+        if unit_ is Any:
+            if isinstance(val, np.ndarray):
+                # input doesn't have a unit, it's just a float-ish type
+                raise ValueError("This needs unit")
+            elif isinstance(val, unit.Quantity):
+                # Redundant cast? Maybe this handles pint vs openff.system.unit?
+                return unit.Quantity(val)
+            else:
+                raise ValueError(
+                    f"Bad input, expected ndarray or pint.Quantity-like, got {type(val)})"
+                )
+        else:
+            unit_ = unit(unit_)
+            if isinstance(val, unit.Quantity):
+                assert unit_.dimensionality == val.dimensionality
+                return val.to(unit_)
+            elif isinstance(val, np.ndarray):
+                return val * unit_
+            elif isinstance(val, list):
+                return np.array(val) * unit_
```

Comment:
```suggestion
            elif isinstance(val, (list, np.ndarray)):
                return val * unit_
```

It's possible that `pint` handles this for us, as it casts `list * Unit` into an array (before it is passed to any other methods)